In [2]:
import pandas as pd
import numpy as np
import datetime as dt
pd.options.display.max_columns=999

# 1.Read Data

In [3]:
df=pd.read_csv('us_equity_daily_price_201308_201908.csv',parse_dates=['date'])
print(df.head())
df.columns

        date   entity_figi ticker       open       high        low      close  \
0 2013-08-01  BBG000C2M952   AAON  14.480000  15.100000  14.360000  15.006667   
1 2013-08-02  BBG000C2M952   AAON  14.966666  15.320000  14.966666  15.260000   
2 2013-08-05  BBG000C2M952   AAON  15.280000  15.473333  15.033334  15.446667   
3 2013-08-06  BBG000C2M952   AAON  15.446667  15.466666  15.320000  15.333333   
4 2013-08-07  BBG000C2M952   AAON  15.333333  15.520000  15.300000  15.466666   

     volume  
0  353700.0  
1  235800.0  
2   90300.0  
3  304800.0  
4  128400.0  


Index(['date', 'entity_figi', 'ticker', 'open', 'high', 'low', 'close',
       'volume'],
      dtype='object')

# 2.Judge Nan

In [4]:
#Sort values
test=df.sort_values(["ticker", "date"]).reset_index(drop=True)
#count nan
count=df.isnull().sum()
print(f'The number of nan is \n {count}')
test["Next_date"] = test.groupby("ticker")["date"].shift(-1)
print(test.head(10))

The number of nan is 
 date             0
entity_figi      0
ticker           0
open           488
high           488
low            488
close          488
volume         488
dtype: int64
        date   entity_figi ticker       open       high        low      close  \
0 2013-08-01  BBG000C2V541      A  32.553650  33.047211  32.417740  32.932762   
1 2013-08-02  BBG000C2V541      A  32.947067  33.268955  32.896996  33.175964   
2 2013-08-05  BBG000C2V541      A  33.090130  33.190273  32.761086  32.861229   
3 2013-08-06  BBG000C2V541      A  32.854076  33.061516  32.625179  32.961372   
4 2013-08-07  BBG000C2V541      A  32.818314  33.233189  32.632332  33.104435   
5 2013-08-08  BBG000C2V541      A  33.261803  33.297569  32.646637  32.796852   
6 2013-08-09  BBG000C2V541      A  32.796852  33.233189  32.768242  33.061516   
7 2013-08-12  BBG000C2V541      A  32.825466  33.268955  32.703861  33.226036   
8 2013-08-13  BBG000C2V541      A  33.226036  33.605148  33.075821  33.562233   
9 

In [5]:
#Show dates of missing data
test[test.open.isna()].date.unique()

array(['2019-08-16T00:00:00.000000000', '2018-11-06T00:00:00.000000000'],
      dtype='datetime64[ns]')

In [6]:
#fill nan
test[(test.open.isna())& (test.date=='2018-11-06T00:00:00.000000000')].index

Int64Index([4237695], dtype='int64')

In [7]:
test[(test.open.isna())& (test.date=='2019-08-16T00:00:00.000000000')].count()

date           487
entity_figi    487
ticker         487
open             0
high             0
low              0
close            0
volume           0
Next_date        0
dtype: int64

# We get 487 2019-08-16 missing data witch is reasonable.  All we need to do is to fill the missing data on 2018-11-06

In [8]:
test.loc[4237695, 'open'] = 3.31
test.loc[4237695, 'high'] = 3.46
test.loc[4237695, 'low'] = 3.30
test.loc[4237695, 'close'] = 3.39
test.loc[4237695, 'volume'] = 1486000
count_new=test.isnull().sum()
count_new

date              0
entity_figi       0
ticker            0
open            487
high            487
low             487
close           487
volume          487
Next_date      4950
dtype: int64

# 3.Gap

In [9]:
test['gap']=test['Next_date']-test.date
print(test.head())
test['gap'] = test['gap'].dt.days
print(test.head())
gap_count=test[test['gap']>3].date.count()
print(f'The total number of gap which larger than 3 is {gap_count}')

        date   entity_figi ticker       open       high        low      close  \
0 2013-08-01  BBG000C2V541      A  32.553650  33.047211  32.417740  32.932762   
1 2013-08-02  BBG000C2V541      A  32.947067  33.268955  32.896996  33.175964   
2 2013-08-05  BBG000C2V541      A  33.090130  33.190273  32.761086  32.861229   
3 2013-08-06  BBG000C2V541      A  32.854076  33.061516  32.625179  32.961372   
4 2013-08-07  BBG000C2V541      A  32.818314  33.233189  32.632332  33.104435   

      volume  Next_date    gap  
0  3827000.0 2013-08-02 1 days  
1  3354000.0 2013-08-05 3 days  
2  2621800.0 2013-08-06 1 days  
3  3188600.0 2013-08-07 1 days  
4  2379500.0 2013-08-08 1 days  
        date   entity_figi ticker       open       high        low      close  \
0 2013-08-01  BBG000C2V541      A  32.553650  33.047211  32.417740  32.932762   
1 2013-08-02  BBG000C2V541      A  32.947067  33.268955  32.896996  33.175964   
2 2013-08-05  BBG000C2V541      A  33.090130  33.190273  32.761086  32.8

In [10]:
test.tail()

date   entity_figi ticker   open   high    low  close    volume  \
5755894 2018-05-16  BBG009J96R40   ZYNE   9.47   9.53  9.070   9.40  154600.0   
5755895 2018-05-17  BBG009J96R40   ZYNE   9.47   9.97  9.350   9.81  247900.0   
5755896 2018-05-18  BBG009J96R40   ZYNE   9.88  10.47  9.870  10.28  285600.0   
5755897 2018-05-21  BBG009J96R40   ZYNE  10.36  10.54  9.621   9.98  292200.0   
5755898 2018-05-22  BBG009J96R40   ZYNE  10.06  10.64  9.901  10.34  261000.0   

         Next_date  gap  
5755894 2018-05-17  1.0  
5755895 2018-05-18  1.0  
5755896 2018-05-21  3.0  
5755897 2018-05-22  1.0  
5755898        NaT  NaN

In [11]:
#Show all information about the company which's gap larger than 3
test_gap_larger_3=test.drop(test[test['gap']<4].index)
test_gap_larger_3.dropna(inplace=True)
test_gap_larger_3

date   entity_figi ticker       open       high        low  \
21      2013-08-30  BBG000C2V541      A  33.540771  33.605148  33.168812   
117     2014-01-17  BBG000C2V541      A  43.512161  43.512161  43.154507   
136     2014-02-14  BBG000C2V541      A  40.879829  41.323318  39.248928   
179     2014-04-17  BBG000C2V541      A  38.440628  39.263233  38.347641   
204     2014-05-23  BBG000C2V541      A  39.806866  40.214592  39.470673   
232     2014-07-03  BBG000C2V541      A  41.623749  41.874107  41.502148   
272     2014-08-29  BBG000C2V541      A  41.008583  41.130184  40.829758   
368     2015-01-16  BBG000C2V541      A  37.830002  38.459999  37.759998   
387     2015-02-13  BBG000C2V541      A  39.970001  40.310001  39.910000   
420     2015-04-02  BBG000C2V541      A  41.470001  42.130001  41.330002   
455     2015-05-22  BBG000C2V541      A  42.270000  42.669998  42.070000   
483     2015-07-02  BBG000C2V541      A  39.360001  39.639999  39.150002   
528     2015-09-04  BBG000C2V541      A  35.040001  35.509998  34.900002   
605     2015-12-24  BBG000C2V541      A  41.790001  42.259998  41.669998   
609     2015-12-31  BBG000C2V541      A  41.900002  42.349998  41.720001   
619     2016-01-15  BBG000C2V541      A  36.669998  37.560001  36.549999   
638     2016-02-12  BBG000C2V541      A  35.840000  36.230000  35.520000   
666     2016-03-24  BBG000C2V541      A  39.160000  39.450001  38.889999   
711     2016-05-27  BBG000C2V541      A  45.549999  45.849998  45.369999   
735     2016-07-01  BBG000C2V541      A  44.360001  44.799999  44.310001   
779     2016-09-02  BBG000C2V541      A  47.230000  47.389999  46.889999   
857     2016-12-23  BBG000C2V541      A  46.080002  46.230000  45.930000   
861     2016-12-30  BBG000C2V541      A  45.759998  45.820000  45.380001   
870     2017-01-13  BBG000C2V541      A  48.599998  48.849998  48.410000   
894     2017-02-17  BBG000C2V541      A  50.650002  51.639999  50.580002   
932     2017-04-13  BBG000C2V541      A  52.430000  52.740002  52.310001   
962     2017-05-26  BBG000C2V541      A  59.230000  59.720001  59.040001   
1030    2017-09-01  BBG000C2V541      A  64.900002  65.180000  64.209999   
1108    2017-12-22  BBG000C2V541      A  67.410004  67.540001  66.860001   
1112    2017-12-29  BBG000C2V541      A  67.500000  67.580002  66.930000   
...            ...           ...    ...        ...        ...        ...   
5754877 2017-02-17  BBG001NFC969     ZX   1.470000   1.480000   1.420000   
5754915 2017-04-13  BBG001NFC969     ZX   1.550000   1.600000   1.550000   
5754945 2017-05-26  BBG001NFC969     ZX   1.750000   1.790000   1.710000   
5755013 2017-09-01  BBG001NFC969     ZX   1.360000   1.360000   1.340000   
5755091 2017-12-22  BBG001NFC969     ZX   1.220000   1.270000   1.210000   
5755095 2017-12-29  BBG001NFC969     ZX   1.380000   1.380000   1.210000   
5755104 2018-01-12  BBG001NFC969     ZX   1.400000   1.430000   1.330000   
5755128 2018-02-16  BBG001NFC969     ZX   1.160000   1.160000   1.150000   
5755156 2018-03-29  BBG001NFC969     ZX   1.040000   1.190000   1.040000   
5755216 2015-09-04  BBG009J96R40   ZYNE  27.930000  29.500000  27.129999   
5755293 2015-12-24  BBG009J96R40   ZYNE  11.490000  12.770000  10.990000   
5755297 2015-12-31  BBG009J96R40   ZYNE   9.610000  10.200000   9.330000   
5755307 2016-01-15  BBG009J96R40   ZYNE   8.270000   8.440000   7.700000   
5755326 2016-02-12  BBG009J96R40   ZYNE   6.280000   7.070000   6.050000   
5755354 2016-03-24  BBG009J96R40   ZYNE   9.500000  10.240000   9.220000   
5755399 2016-05-27  BBG009J96R40   ZYNE   7.640000   7.900000   7.310000   
5755423 2016-07-01  BBG009J96R40   ZYNE   6.970000   7.176000   6.820000   
5755467 2016-09-02  BBG009J96R40   ZYNE   8.520000   8.680000   8.251000   
5755545 2016-12-23  BBG009J96R40   ZYNE  14.850000  16.000000  14.850000   
5755549 2016-12-30  BBG009J96R40   ZYNE  15.900000  15.900000  15.400000   
5755558 2017-01-13  BBG009J96R40   ZYNE  19.250000  19.438000 

In [12]:
#Show the maximum gap days 
max_gap=test_gap_larger_3.pivot_table(index='ticker',values='gap',aggfunc=max)
print(f'max_gap is: \n {max_gap}')
#Show the largest gap value
max_gap_value=test_gap_larger_3.gap.max()
print(f'max gap value is: \n {max_gap_value}')

max_gap is: 
          gap
ticker      
A        4.0
AAAP     7.0
AAC     17.0
AAL      4.0
AAME     4.0
AAN      4.0
AAOI     4.0
AAON     4.0
AAP      4.0
AAPL     4.0
AAT      4.0
AAV      4.0
AAWW     4.0
AAXJ     4.0
AB       4.0
ABAC     4.0
ABAX     4.0
ABB      4.0
ABBV     4.0
ABC      4.0
ABCB     4.0
ABCD     4.0
ABDC     4.0
ABEO     4.0
ABEV     4.0
ABG      4.0
ABIO     4.0
ABM      4.0
ABMD     4.0
ABR      4.0
...      ...
YY       4.0
Z        4.0
ZAGG     4.0
ZAIS     4.0
ZAYO     4.0
ZAZA     4.0
ZBH     57.0
ZBIO     4.0
ZBRA     4.0
ZEL     82.0
ZEN      4.0
ZEUS     4.0
ZF       4.0
ZFGN     4.0
ZG       4.0
ZGNX     4.0
ZION     4.0
ZIOP     4.0
ZIV      4.0
ZIXI     4.0
ZN       4.0
ZNGA     4.0
ZNH      4.0
ZOES     4.0
ZSAN     4.0
ZTR      4.0
ZTS      4.0
ZUMZ     4.0
ZX       4.0
ZYNE     4.0

[4904 rows x 1 columns]
max gap value is: 
 783.0


In [13]:
#check given company's gap days
test_gap_larger_3.set_index(['ticker', 'date']).loc['A'].gap

date
2013-08-30    4.0
2014-01-17    4.0
2014-02-14    4.0
2014-04-17    4.0
2014-05-23    4.0
2014-07-03    4.0
2014-08-29    4.0
2015-01-16    4.0
2015-02-13    4.0
2015-04-02    4.0
2015-05-22    4.0
2015-07-02    4.0
2015-09-04    4.0
2015-12-24    4.0
2015-12-31    4.0
2016-01-15    4.0
2016-02-12    4.0
2016-03-24    4.0
2016-05-27    4.0
2016-07-01    4.0
2016-09-02    4.0
2016-12-23    4.0
2016-12-30    4.0
2017-01-13    4.0
2017-02-17    4.0
2017-04-13    4.0
2017-05-26    4.0
2017-09-01    4.0
2017-12-22    4.0
2017-12-29    4.0
2018-01-12    4.0
2018-02-16    4.0
2018-03-29    4.0
2018-05-25    4.0
2018-08-31    4.0
2019-01-18    4.0
2019-02-15    4.0
2019-04-18    4.0
2019-05-24    4.0
Name: gap, dtype: float64

# 4.Add data of prior dates

In [14]:
#arrays=[list(test.ticker),list(range(5755898))]
#tuples=list(zip(*arrays))
#index = pd.MultiIndex.from_tuples(tuples, names=['ticker', 'ID'])
#index
#test_new=test.set_index(index)

In [15]:
for i in np.arange(1,6,1):
    test[f'{i}d_prior_open']=pd.Series(test.groupby("ticker")["open"].shift(i))
    test[f'{i}d_prior_high']=pd.Series(test.groupby("ticker")["high"].shift(i))
    test[f'{i}d_prior_low']=pd.Series(test.groupby("ticker")["low"].shift(i))
    test[f'{i}d_prior_close']=pd.Series(test.groupby("ticker")["close"].shift(i))
    test[f'{i}d_prior_volume']=pd.Series(test.groupby("ticker")["volume"].shift(i))
for i in np.arange(-1,-6,-1):
    test[f'{i}d_back_open']=pd.Series(test.groupby("ticker")["open"].shift(i))
    test[f'{i}d_back_high']=pd.Series(test.groupby("ticker")["high"].shift(i))
    test[f'{i}d_back_low']=pd.Series(test.groupby("ticker")["low"].shift(i))
    test[f'{i}d_back_close']=pd.Series(test.groupby("ticker")["close"].shift(i))
    test[f'{i}d_back_volume']=pd.Series(test.groupby("ticker")["volume"].shift(i))

In [16]:
test.head(10)

date   entity_figi ticker       open       high        low      close  \
0 2013-08-01  BBG000C2V541      A  32.553650  33.047211  32.417740  32.932762   
1 2013-08-02  BBG000C2V541      A  32.947067  33.268955  32.896996  33.175964   
2 2013-08-05  BBG000C2V541      A  33.090130  33.190273  32.761086  32.861229   
3 2013-08-06  BBG000C2V541      A  32.854076  33.061516  32.625179  32.961372   
4 2013-08-07  BBG000C2V541      A  32.818314  33.233189  32.632332  33.104435   
5 2013-08-08  BBG000C2V541      A  33.261803  33.297569  32.646637  32.796852   
6 2013-08-09  BBG000C2V541      A  32.796852  33.233189  32.768242  33.061516   
7 2013-08-12  BBG000C2V541      A  32.825466  33.268955  32.703861  33.226036   
8 2013-08-13  BBG000C2V541      A  33.226036  33.605148  33.075821  33.562233   
9 2013-08-14  BBG000C2V541      A  33.454937  33.619457  33.168812  33.268955   

      volume  Next_date  gap  1d_prior_open  1d_prior_high  1d_prior_low  \
0  3827000.0 2013-08-02  1.0            NaN            NaN           NaN   
1  3354000.0 2013-08-05  3.0      32.553650      33.047211     32.417740   
2  2621800.0 2013-08-06  1.0      32.947067      33.268955     32.896996   
3  3188600.0 2013-08-07  1.0      33.090130      33.190273     32.761086   
4  2379500.0 2013-08-08  1.0      32.854076      33.061516     32.625179   
5  2379900.0 2013-08-09  1.0      32.818314      33.233189     32.632332   
6  2737800.0 2013-08-12  3.0      33.261803      33.297569     32.646637   
7  1743500.0 2013-08-13  1.0      32.796852      33.233189     32.768242   
8  2905300.0 2013-08-14  1.0      32.825466      33.268955     32.703861   
9  2778500.0 2013-08-15  1.0      33.226036      33.605148     33.075821   

   1d_prior_close  1d_prior_volume  2d_prior_open  2d_prior_high  \
0             NaN              NaN            NaN            NaN   
1       32.932762        3827000.0            NaN            NaN   
2       33.175964        3354000.0      32.553650      33.047211   
3       32.861229        2621800.0      32.947067      33.268955   
4       32.961372        3188600.0      33.090130      33.190273   
5       33.104435        2379500.0      32.854076      33.061516   
6       32.796852        2379900.0      32.818314      33.233189   
7       33.061516        2737800.0      33.261803      33.297569   
8       33.226036        1743500.0      32.796852      33.233189   
9       33.562233        2905300.0      32.825466      33.268955   

   2d_prior_low  2d_prior_close  2d_prior_volume  3d_prior_open  \
0           NaN             NaN              NaN            NaN   
1           NaN             NaN              NaN            NaN   
2     32.417740       32.932762        3827000.0            NaN   
3     32.896996       33.175964        3354000.0      32.553650   
4     32.761086       32.861229        2621800.0      32.947067   
5     32.625179       32.961372        3188600.0      33.090130   
6     32.632332       33.104435        2379500.0      32.854076   
7     32.646637       32.796852        2379900.0      32.818314   
8     32.768242       33.061516        2737800.0      33.261803   
9     32.703861       33.226036        1743500.0      32.796852   

   3d_prior_high  3d_prior_low  3d_prior_close  3d_prior_volume  \
0            NaN           NaN             NaN              NaN   
1            NaN           NaN             NaN              NaN   
2            NaN           NaN             NaN              NaN   
3      33.047211     32.417740       32.932762        3827000.0   
4      33.268955     32.896996       33.175964        3354000.0   
5      33.190273     32.761086       32.861229        2621800.0   
6      33.061516     32.625179       32.961372        3188600.0   
7      33.233189     32.632332       33.104435        2379500.0   
8      33.297569     32.646637       32.796852        2379900.0   
9      33.233189     32.768242       33.061516        2737800.0   

   4d_prior_open  4d_prior_high  4d_prior_low  4d_prio

In [19]:
test.to_csv("finished_us_equity_daily_price_201308_201908.csv", index=False)